### PR TITLE
[BEAM-3376] Adds possibility to customize of default_value evaluation in CombineFn.

### PR DESCRIPTION
**Pull request by tvalentyn**
This change makes it possible to customize default value evaluation in CombineFn to match Java SDK, see: 
https://github.com/apache/beam/blob/f639a7e760dcbe82f2d0236f715e964dd746931d/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Combine.java#L461

If necessary users could to skip CombineFn.setup and CombineFn.teardown during calculation of the default value during initial pipeline expansion, see discussion in upcoming https://github.com/apache/beam/pull/13048#discussion_r503585072. 

Also adds some rewording/explanation for default_value usage following previous PRs (#12074, #3984).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by tvalentyn** [comment]
R: @kamilwu   
cc: @aaltay @robertwb  @pabloem @InigoSJ  who may have feedback.

**Comment by InigoSJ** [comment]
> R: @kamilwu
> cc: @aaltay @robertwb @pabloem @InigoSJ who may have feedback.

It looks so much better than `CombineWithoutDefaults`, thanks for that.

File: sdks/python/apache_beam/transforms/core.py
Comment (kamilwu) [inline sdks/python/apache_beam/transforms/core.py:908]: nit: Default

File: sdks/python/apache_beam/transforms/core.py
Comment (kamilwu) [inline sdks/python/apache_beam/transforms/core.py:911]: The user can provide a combiner that has nested combiners, e.g. `TupleCombineFn`, so we should make a deep copy instead.

**Comment by aaltay** [comment]
Renaming `CombinerWithoutDefaults` would be backwards incompatible. Users needs to change their potentially and I think we can avoid that part of the change. Rest of the change LGTM.

**Comment by robertwb** [comment]
Note that one can always override `CombineFn.apply()`, so this seems more about setup/teardown, right? 

The disadvantage of a special method means that the semantics are unclear of whether one is allowed to call apply() with an empty iterable... or is one forced to check if the iterable is empty to choose to call apply or default_value? I'm thinking about other places CombineFns may be used used, for example, in CombiningStates. 

If the goal is just to avoid calling setup/teardown in the main program, we can rewrite the `InjectDefault` inside `CombineGlobally` to take an iterable side input and invoke `apply([])` manually (with the setup and teardown) iff the iterable is empty.

**Comment by tvalentyn** [comment]
> Renaming `CombinerWithoutDefaults` would be backwards incompatible. Users needs to change their potentially and I think we can avoid that part of the change. Rest of the change LGTM.

It's a fairly recent introduction, see: #12074, and not included in https://github.com/apache/beam/blob/c3477a8eadf3fef6c173166bb1f720ed6c8ffe42/sdks/python/apache_beam/transforms/combiners.py#L56.

Do you still think renaming would be a concern? If so I can add an alias.

**Comment by tvalentyn** [comment]
> Note that one can always override CombineFn.apply(), so this seems more about setup/teardown, right?
Yes, setup/teardown was primary motivation for writing this up. 

> The disadvantage of a special method means that the semantics are unclear of whether one is allowed to call apply() with an empty iterable... or is one forced to check if the iterable is empty to choose to call apply or default_value? I'm thinking about other places CombineFns may be used used, for example, in CombiningStates.

I can clarify in the api that this is optional, and defaults to empty apply() call if we decide to proceed as is.

> If the goal is just to avoid calling setup/teardown in the main program
In your opinion, is calling setup/teardown in main program a concern at all? Perhaps I am overthinking this.

> we can rewrite the InjectDefault inside CombineGlobally to take an iterable side input and invoke apply([]) manually (with the setup and teardown) iff the iterable is empty.
So in this case evaluation of the default will happen after job submission?  I admit I don't fully follow the idea, so if writing a prototype of this idea is faster than explaining how to do it right to myself of Kamil, we'd appreciate a draft; thanks!

**Comment by robertwb** [comment]
Here's what I was thinking of in terms of deferring the call to `apply()` to workers: https://github.com/apache/beam/pull/13132

I am still generally opposed to having a separate `default_value` method because it opens up ambiguity of when it is called instead of `apply([])` as well as surprising behavior if the two do not agree.

**Comment by tvalentyn** [comment]
> Here's what I was thinking of in terms of deferring the call to apply() to workers: #13132
> 
> I am still generally opposed to having a separate default_value method because it opens up ambiguity of when it is called instead of apply([]) as well as surprising behavior if the two do not agree.

Thanks for your input. I reverted the default() changes and only keep the clarification/clean-up portion.

**Comment by TheNeuralBit** [comment]
@robertwb I think this is still waiting for your review

**Review by robertwb — APPROVED** [review]
LGTM if we don't think the renaming will cause a problem. (Otherwise we could add an alias.)

**Pull request closed without merging**
